### PR TITLE
Fix the file name already exists on the server when uploading the file

### PR DESCRIPTION
**Pull request by phoon**

### CODE REVIEW

**Comment by cpg** [comment]
This issue [reminded me of this](https://i.redd.it/gs7ci2yawtj21.jpg)

In any case, I am not sure I like the idea of a random string. Especially at the beginning of the file name. This can have the effect of the user not seeing the files listed together (typically the default is often list files alphabetically, so they would be potentially far apart).

I feel that that a little forward thinking can be useful here.

What other options do we have?

0) random string at the beginning (what you proposed)
1) random string at the end (preserving the extension), i.e. XYZ.foo XYZ-hj8d3y.foo
2) like the above image .. have a (1) and (2) and (3), etc. when the file exists: XYZ (1).foo and XYZ (2).foo, etc. (the separator being a space). This involves one more loop to check if other file names exist and incrementing the counter (though even random ought to check if the new filename exists, remote as it may be). 
3) a date/timestamp, e.g. if XYZ.foo exists -- XYZ-20190304-1420.foo
4) do a little what Dropbox does -- e.g. XYZ (conflicted copy 2019-03-04).foo

Option 2 has the advantage of being widely known by many of the average users at large, since Windows uses this.

I kind of like 3 and 4.

Note: for any extensions that are added at the end, the extension must be preserved, for (I hope) obvious reasons

Any thoughts, guys?

**Comment by phoon** [comment]
the new commit a72f804 implements use a timestamp suffix to rename the file, i think this is a good way to solve the problem, user can use the timestamp to determine the update of the file.

**Review by csoni111 — APPROVED** [review]
Looks good! can be merged @cpg

**Comment by csoni111** [comment]
About matching md5 if the file exists (as @iPeven mentioned in #17), I think we should include that too in this pr. It will ensure user doesn't accidentally upload a file twice.

**Comment by phoon** [comment]
Actually, SHA is more secure than MD5 on hash collisions. What bothers me a lot is it will takes a long time to calculate the hash if the file is too large. And I have no idea about where to store it at now.

**Comment by csoni111** [comment]
I believe MD5 is good enough to check file duplicacy. Most of the file duplicate checkers use MD5. This [answer](https://stackoverflow.com/a/24023524/6352482) on SO gives a rough probability of finding same hash for two different files which is very very low. 
As for storing it, we can calculate it at run time, no need to store it as it will happen for only those uploads whose file name and size matches first with already existing files.

**Comment by phoon** [comment]
Done!The processing flow is as follows:
```
 upload a file ---(not exists)---> create and write directly
             |
              ---(exists)---> calculate the MD5 ---(MD5 is different)---> Rename the uploaded one, create&write
                                               |
                                                ---(Same MD5)---> ignore the uploaded file



```

File: src/fs/service.go [hunk 1]
```diff
@@ -650,3 +677,60 @@ func (service *MercuryFsService) uploadFile(writer http.ResponseWriter, request
...
+		//It could be slash after the system escaped it automatically.
+		strings.HasPrefix(f, "⁄") ||
+		strings.HasPrefix(f, " ") {
+		return false
+	}
+	return true
+}
+
+//calculate the md5
+func CalMD5(input io.Reader) string {
```
Comment (csoni111) [inline src/fs/service.go]: Keep the function name in camelCase.

File: src/fs/service.go [hunk 2]
```diff
@@ -650,3 +677,60 @@ func (service *MercuryFsService) uploadFile(writer http.ResponseWriter, request
...
+}
+
+//check if the file name is valid
+func validFilename(f string) bool {
+	if strings.HasPrefix(f, ".") ||
+		strings.HasPrefix(f, "/") ||
+		//It could be slash after the system escaped it automatically.
+		strings.HasPrefix(f, "⁄") ||
+		strings.HasPrefix(f, " ") {
+		return false
```
Comment (csoni111) [inline src/fs/service.go]: Only these many checks are not sufficient, there may be system reserved keywords and other things like max filename length depending on the OS. Check this answer: https://stackoverflow.com/a/35240286/6352482 , the second approach can be taken here.
Ref: https://en.wikipedia.org/wiki/Filename (contains list of reserved keywords, filename length restrictions for different os)

File: src/fs/service.go [hunk 3]
```diff
@@ -650,3 +677,60 @@ func (service *MercuryFsService) uploadFile(writer http.ResponseWriter, request
 
 	return
 }
+
+// rename the given file path
+func renameFile(p string) string {
```
Comment (csoni111) [inline src/fs/service.go]: could move all these functions related to file handling in a new go file say `file_upload_utils.go`.

File: src/fs/service.go [hunk 4]
```diff
@@ -650,3 +677,60 @@ func (service *MercuryFsService) uploadFile(writer http.ResponseWriter, request
...
+	timeStamp := time.Now().Format("20060102-1504")
+
+	return baseName + "-" + timeStamp + ext
+}
+
+/*
+status:
+0 => file not exists,
+1 => file exists,
+2 => file exists and have the same md5
```
Comment (csoni111) [inline src/fs/service.go]: You can make an enum for these. Make a custom int type and use iota for assigning them values automatically.
Ref: https://blog.learngoprogramming.com/golang-const-type-enums-iota-bc4befd096d3

File: src/fs/service.go [hunk 5]
```diff
@@ -626,17 +629,41 @@ func (service *MercuryFsService) uploadFile(writer http.ResponseWriter, request
...
+			writer.WriteHeader(http.StatusUnsupportedMediaType)
+			service.debugInfo.requestServed(int64(0))
+			log("\"POST %s\" 415 0 \"%s\"", query, ua)
+			return
+		}
+
 		fullPath, _ := service.fullPathToFile(share, path+"/"+handler.Filename)
 
-		f, err := os.OpenFile(fullPath, os.O_WRONLY|os.O_CREATE, 0644)
+		md5 := CalMD5(file)
```
Comment (csoni111) [inline src/fs/service.go]: This should not be calculated here, calculate MD5 for the uploaded file only if another file exists with the same name.

**Comment by phoon** [comment]
@csoni111  Thank you for your code reviewing, I apologize for my lack of project experience.

**Pull request merged**